File: backend/services/post-ts-koa-mongo-fdm/src/admin-login/impl.ts
import { AdminId } from "../common";
import { Workflow } from ".";
import { Deps } from "../common/utils";
import { WithId } from "mongodb";
import * as db from "../common/utils/db";
import { errors } from "./api";
import * as runtypes from "runtypes";
import { PasswordVerifier } from "../common/utils/password";

export class WorkflowImpl extends Workflow {
    private static readonly adminHasPassword = runtypes.Record({
        encryptedPassword: runtypes.String,
    });
    async getPasswordVerifier(id: AdminId): Promise<PasswordVerifier> {
        const oid = db.tryParseId(id);
        if (oid === undefined) {
            throw errors.idOrPasswordIncorrect();
        }
        const rec: WithId<unknown> | null = await this.deps.mongoDb.collection(db.admins).findOne({ _id: oid });
        if (rec === null) {
            throw this.errors.idOrPasswordIncorrect();
        }
        db.validate(WorkflowImpl.adminHasPassword, db.admins, rec);

        return new db.BCryptVerifier(rec.encryptedPassword);
    }

    readonly errors = errors;
    constructor(private deps: Deps) { super(); }
}

File: backend/services/post-kt-vertx-fdm/src/main/kotlin/me/jason5lee/post_kt_vertx_fdm/admin_login/impl.kt
package me.jason5lee.post_kt_vertx_fdm.admin_login

import com.mongodb.client.model.Projections
import me.jason5lee.post_kt_vertx_fdm.common.AdminId
import me.jason5lee.post_kt_vertx_fdm.common.Password
import me.jason5lee.post_kt_vertx_fdm.common.utils.Db
import me.jason5lee.post_kt_vertx_fdm.common.utils.Deps

class WorkflowImpl(private val deps: Deps) : Workflow(), FailuresImpl {
    override suspend fun verifyPassword(id: AdminId, password: Password) {
        val admin = Db.findById(
            db = deps.mongoDb,
            collection = Db.admins,
            id = Db.toObjectIdOrNull(id.value) ?: throw idOrPasswordIncorrect(),
            projection = Projections.include("encryptedPassword")
        ) ?: throw idOrPasswordIncorrect()
        val encryptedPassword = admin.get<String>("encryptedPassword")
        if (!password.verify(Db.BCrypt.Verifier(encryptedPassword))) {
            throw idOrPasswordIncorrect()
        }
    }
}
